lib/Dialect/TTNN/Analysis/BFInterleavedPolicy.cpp

```diff
@@ -0,0 +1,207 @@
+
+          if (currentL1Usage + opOutputL1Usage <= getAvailableL1CacheSize()) {
+            allocOfL1Mem = opOutputL1Usage;
+            opBufferType = BufferType::L1;
+          }
+        }
+
+        // Check if the scheduling of the op is consumes the least amount of L1
+        // memory among all the scheduleable ops.
+        //
+        changeInL1Usage = allocOfL1Mem - deallocOfL1Mem;
+        if (changeInL1Usage < minimalChangeInL1Usage) {
+          nextOpForScheduling = op;
+          nextOpForSchedulingBufferType = opBufferType;
+          minimalChangeInL1Usage = changeInL1Usage;
+        }
+      }
+
+      // In case we picked the L1 layout for the nextOpForScheduling, we need
+      // to add the OpL1MemSpec to the L1ChainConfig and update the state of L1
+      // memory.
+      //
+      if (nextOpForSchedulingBufferType == BufferType::L1) {
+
+        // Construct OpL1MemSpec for the nextOpForScheduling.
+        //
+        OpL1MemSpec opL1MemSpec;
+        opL1MemSpec.op = nextOpForScheduling;
+        opL1MemSpec.layout = getL1InterleavedLayout(nextOpForScheduling);
+        l1ChainConfigs->back().addOpL1MemSpec(opL1MemSpec);
+
+        // Update the state of L1 memory by allocating the nextOpForScheduling's
+        // output tensor in L1 memory.
+        //
+        uint64_t numOfUsers = std::distance(nextOpForScheduling->user_begin(),
+                                            nextOpForScheduling->user_end());
+        currentL1UsagePerOp[nextOpForScheduling].l1MemUsagePerUser =
+            utils::getOpOutputL1Usage(nextOpForScheduling, opL1MemSpec.layout,
+                                      deviceAttr);
+        currentL1UsagePerOp[nextOpForScheduling].numOfUnscheduledUsers =
+            numOfUsers;
+        currentL1Usage +=
+            currentL1UsagePerOp[nextOpForScheduling].l1MemUsagePerUser;
+      }
+
+      // Update the state of L1 memory.
+      //
+      walkOnAnalyzableOperands(nextOpForScheduling, [&](Operation *operandOp) {
+        if (currentL1UsagePerOp.count(operandOp)) {
+          currentL1UsagePerOp[operandOp].numOfUnscheduledUsers -= 1;
+          if (currentL1UsagePerOp[operandOp].numOfUnscheduledUsers == 0) {
+            // Only once we scheduled all the users of the operandOp, we can
+            // free its output tensor from L1 memory.
+            //
+            currentL1Usage -= currentL1UsagePerOp[operandOp].l1MemUsagePerUser;
+            currentL1UsagePerOp.erase(operandOp);
+          }
+        }
+      });
+
+      // Schedule the nextOpForScheduling and update currentL1Usage.
+      //
+      scheduler.scheduleOp(nextOpForScheduling);
+    }
+
+    assert(currentL1Usage == 0);
+    assert(currentL1UsagePerOp.size() == 0);
+
+    (*schedule)[func] = scheduler.getSchedule();
+
+    // Build, Resolve and Complete all L1ChainConfigs.
+    //
+    for (L1ChainConfig &l1ChainConfig : *l1ChainConfigs) {
+      l1ChainConfig.build();
```

Comment:
based on logic above I would say its possible to do this per chainconfig on the go in the main loop.

Reply:
Are you referring to creating multiple l1ChainConfigs on the go?

If so, these would be subgraphs. If there is no separate chain build / resolve etc. logic, does doing this provide any value for this specific policy implementation?

Reply:
We are creating single chain per func? I was referring that even for that process we can have meaningful points when build, resolve and complete phase are performed. This is not a hard requirement of course.

Reply:
I only have one l1ChainConfig. That's the reason why we need to wait until all the ops are analysed.